src/Rules/Functions/PrintfArrayParametersRule.php

```diff
@@ -0,0 +1,184 @@
+		$name = $functionReflection->getName();
+		if (!in_array($name, ['vprintf', 'vsprintf'], true)) {
+			return [];
+		}
+
+		$args = $node->getArgs();
+		$argsCount = count($args);
+		if ($argsCount < 1) {
+			return []; // caught by CallToFunctionParametersRule
+		}
+
+		$formatArgType = $scope->getType($args[0]->value);
+		$placeHoldersCounts = [];
+		foreach ($formatArgType->getConstantStrings() as $formatString) {
+			$format = $formatString->getValue();
+
+			$placeHoldersCounts[] = $this->printfHelper->getPrintfPlaceholdersCount($format);
+		}
+
+		if ($placeHoldersCounts === []) {
+			return [];
+		}
+
+		$minCount = min($placeHoldersCounts);
+		$maxCount = max($placeHoldersCounts);
+		if ($minCount === $maxCount) {
+			$placeHoldersCount = new ConstantIntegerType($minCount);
+		} else {
+			$placeHoldersCount = IntegerRangeType::fromInterval($minCount, $maxCount);
+		}
+
+		$formatArgsCounts = [];
+		if (isset($args[1])) {
+			$formatArgsType = $scope->getType($args[1]->value);
+
+			$constantArrays = $formatArgsType->getConstantArrays();
+			foreach ($constantArrays as $constantArray) {
+				$formatArgsCounts[] = $constantArray->getArraySize();
+			}
+
+			if ($constantArrays === []) {
+				$formatArgsCounts[] = $formatArgsType->getArraySize();
+			}
+		}
+
+		if ($formatArgsCounts === []) {
+			$formatArgsCount = new ConstantIntegerType(0);
+		} else {
+			$formatArgsCount = TypeCombinator::union(...$formatArgsCounts);
+		}
+
+		if (!$this->placeholdersMatchesArgsCount($placeHoldersCount, $formatArgsCount)) {
+
+			if ($placeHoldersCount instanceof IntegerRangeType) {
+				$placeholders = $this->getIntegerRangeAsString($placeHoldersCount);
+				$singlePlaceholder = false;
+			} elseif ($placeHoldersCount instanceof ConstantIntegerType) {
+				$placeholders = $placeHoldersCount->getValue();
+				$singlePlaceholder = $placeholders === 1;
+			} else {
+				throw new ShouldNotHappenException();
+			}
+
+			if ($formatArgsCount instanceof IntegerRangeType) {
+				$values = $this->getIntegerRangeAsString($formatArgsCount);
+				$singleValue = false;
+			} elseif ($formatArgsCount instanceof ConstantIntegerType) {
+				$values = $formatArgsCount->getValue();
+				$singleValue = $values === 1;
+			} else {
+				throw new ShouldNotHappenException();
+			}
+
+			return [
+				RuleErrorBuilder::message(sprintf(
+					sprintf(
+						'%s, %s.',
+						$singlePlaceholder ? 'Call to %s contains %d placeholder' : 'Call to %s contains %s placeholders',
+						$singleValue ? '%d value given' : '%s values given',
+					),
+					$name,
+					$placeholders,
+					$values,
+				))->identifier(sprintf('argument.%s', $name))->build(),
+			];
+		}
+
+		return [];
+	}
+
+	private function placeholdersMatchesArgsCount(Type $placeHoldersCount, Type $formatArgsCount): bool
+	{
+		if ($placeHoldersCount instanceof ConstantIntegerType && $formatArgsCount instanceof ConstantIntegerType) {
```

Comment:
This kind of structure makes it hard to see if all possible cases are handled.

1) Please add more specific parameter types, we know these can only be ConstantIntegerType/IntegerRangeType.
2) Do `if ($placeHoldersCount instanceof ConstantIntegerType) { ... }` and then handle all possible conditions inside. Then do `else` and handle all possible conditions again. Do not compound conditions in the method body on the top level with `&&`. It makes it hard to see for example whether `$a` with ConstantIntegerType and `$b` with IntegerRangeType is handled...

Reply:
I can't because `IntegerRangeType::fromInterval(..)` can return any Type `NeverType|ConstantIntegerType|IntegerType|IntegerRangeType`